src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs

```diff
@@ -1297,6 +1313,100 @@ private BoundExpression RewriteDecimalConversion(BoundConversion oldNode, CSharp
+            // b. currentMethod.ContainingType or/and its ancestors could be generic, making the container type scoped could possibly cause multiple instances at runtime
+
+            // However, there are cases when targetMethod or/and delegateType being private or only visible inside currentMethod.ContainingType by code.
+            // It's not a problem that targetMethod being not visible, but delegateType could. Even it seems to work, it's not worth the risk.
+            if (IsVisibleAtModuleScope(delegateType))
+            {
+                // If the currentMethod and it's containing type and it's ancestors all the way up are concrete,
+                // then the delegate type and target method should be fully concrete.
+                if (currentMethod.Arity == 0 && !currentMethod.ContainingType.IsGenericType)
+                {
+                    return DelegateCacheContainerKind.ModuleScopedConcrete;
+                }
+
+                // Although the currentMethod or it's containing types may be generic, the delegateType and targetMethod may not. Let's find out.
+                var fullyConcreteChecker = FullyConcreteChecker.Instance;
+                if (fullyConcreteChecker.Visit(delegateType) && fullyConcreteChecker.Visit(targetMethod))
+                {
+                    return DelegateCacheContainerKind.ModuleScopedConcrete;
+                }
+            }
+
+            // Now the conversion could involve generics. All the possible type parameters that act as type arguments
+            // needed to construct the delegateType or targetMethod come from either the current method or it's containing types
+
+            // So obviously,
+            if (currentMethod.Arity == 0)
+            {
+                return DelegateCacheContainerKind.TypeScopedConcrete;
+            }
+
+            // Then we need to really check if currentMethod.TypeParameters are involved.
+            var typeParams = currentMethod.TypeParameters;
+            var typeParamUsageChecker = TypeParameterUsageChecker.Instance;
+            if (typeParamUsageChecker.Visit(delegateType, typeParams) || typeParamUsageChecker.Visit(targetMethod, typeParams))
+            {
+                // If so, we have to "mock" the type parameters of the currentMethod as type parameters of the cache container.
+                // So we can later use them to define the delegate type of the fields of the cache container.
+                // That being said, we need a generic container here.
+                return DelegateCacheContainerKind.MethodScopedGeneric;
+            }
+
+            // If not, we can just use the delegateType for the cache field of the container as is.
+            return DelegateCacheContainerKind.TypeScopedConcrete;
+        }
+
+        private static bool IsVisibleAtModuleScope(NamedTypeSymbol type)
```

Comment:
Rather than inventing yet another way of checking accessibility, I think we should use an existing helper in Microsoft.CodeAnalysis.CSharp.AccessCheck
```
        /// <summary>
        /// Checks if 'symbol' is accessible from within assembly 'within'.  
        /// </summary>
        public static bool IsSymbolAccessible(
            Symbol symbol,
            AssemblySymbol within,
            ref HashSet<DiagnosticInfo> useSiteDiagnostics)
        {
            bool failedThroughTypeCheck;
            return IsSymbolAccessibleCore(symbol, within, null, out failedThroughTypeCheck, within.DeclaringCompilation, ref useSiteDiagnostics);
        }
```